src/vision/status_quo/alan_tries_a_socket_sink.md

```diff
@@ -0,0 +1,106 @@
+# 😱 Status quo stories: Alan tries using a socket Sink
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today.
+
+If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!
+
+## The story
+
+Alan is working on a project that uses an alternative popular async framework. He finds the new async framework less familiar, and as he reads through their GitHub issues, he's saddened by the project maintainer's unwelcoming attitude towards their community. He asks his teammates if they can switch to the one he's more familiar with and feels more welcome in, but the teammates are understandably resistant to changing something that important just for one feature. So, he works hard to find workarounds to accomplish his team's goals.
+
+One of the goals is to switch from a WebSocket implementation using raw TCP sockets to one managed behind an HTTP server library, so both HTTP and WebSocket commands can be forwarded to a transport-agnostic RPC server. He finds an HTTP server that's similar to one he's used to using with the other async framework, and a WebSocket middleware library that goes with it.
+
+However, as he's working, Alan encounters a situation where the socket needs to be written to within an async thread, and the traits just aren't working. He wants to split the stream into a sender and receiver:
+
+```rust
+use futures::{SinkExt, StreamExt};
+use async_std::sync::{Arc, Mutex};
+use log::{debug, info, warn};
+
+/// In the connection handler:
+let (ws_sender, mut ws_receiver) = ws_stream.split();
+let ws_sender = Arc::new(Mutex::new(ws_sender));
+
+while let Some(msg) = ws_receiver.next().await {
+    debug!("Received new WS RPC message: {:?}", msg);
+
+    // Echo the request:
+    match ws_sender.lock().await.send_string(msg).await {
+        Ok(_) => info!("New WS data sent."),
+        Err(_) => warn!("WS connection closed."),
+    };
+}
+```
+
+This is necessary because both sides of the stream are stateful iterators, and if they're kept as one, the lock on the receiver can't be released within the loop. He's seen this pattern used in other projects in the Rust community, but is frustrated to find that the `Sink` trait wasn't implemented in the WebSockets middleware library he's using.
```

Comment:
I'm not sure how better to explain it...

Reply:
Like, I literally don't see a lock on `ws_receiver`, so that is still confusing to me. =) I guess the point is that `ws_receiver` can only produce one message at a time, but you want to introduce some parallelism after they are produced? (This is what I tried to note above, but elaborated)

Reply:
Yep, that's by design! `ws_receiver` in both cases is single-threaded, and thus does not need to be sent across multiple threads nor is it necessary to lock.